### PR TITLE
Lvnod 57 create registration component

### PR DESCRIPTION
**Pull request by yurii-kokhalevych**
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/browse/LVNOD-57)
* [JIRA subtask](https://ssu-jira.softserveinc.com/browse/LVNOD-XXX)

## Code reviewers

- [x] @yarvanya
- [x] @AlbaMontis
- [x] @avoytovich

### Second Level Review

- [ ] @lhalam
- [ ] @4asof4ik

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo
## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: app/src/containers/Signup/Signup.js [hunk 1]
```diff
@@ -1,9 +1,142 @@
-import React from 'react';
+import React, { Component } from 'react';
+import { FormGroup, ControlLabel, FormControl, Button, ButtonToolbar, Form, HelpBlock } from 'react-bootstrap';
+import { browserHistory } from 'react-router';
+import style from './Signup.scss';
 
-export default class Signup extends React.Component {
-  render () {
+let request = require('superagent');
```
Comment (avoytovich) [inline app/src/containers/Signup/Signup.js]: import request from 'superagent';

**Review by yarvanya — DISMISSED** [review]

File: app/src/containers/Signup/Signup.js [hunk 2]
```diff
@@ -1,9 +1,142 @@
...
+let user = {firstName: 'Ivan', lastName: 'Popov', email: 'rrr@ukr.net'};
+
+const FieldGroup = ({className, label, inErrors, ...props}) => {
```
Comment (avoytovich) [inline app/src/containers/Signup/Signup.js]: Normally you would define a React component as a plain JavaScript class:
class Greeting extends React.Component {
  render() {
    return <h1>Hello, {this.props.name}</h1>;
  }
}

File: app/src/containers/Signup/Signup.js [hunk 3]
```diff
@@ -1,9 +1,142 @@
-import React from 'react';
+import React, { Component } from 'react';
```
Comment (avoytovich) [inline app/src/containers/Signup/Signup.js]: why you use { Component } ???

Comment (yurii-kokhalevych) [reply app/src/containers/Signup/Signup.js]: export default class Signup extends Component

File: app/src/containers/Signup/Signup.js [hunk 4]
```diff
@@ -1,9 +1,142 @@
...
+import { FormGroup, ControlLabel, FormControl, Button, ButtonToolbar, Form, HelpBlock } from 'react-bootstrap';
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Do you use ButtonToolbar somewhere?

File: app/src/containers/Signup/Signup.js [hunk 5]
```diff
@@ -1,9 +1,142 @@
...
+    const pass = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[$@$!%*?&])[A-Za-z\d$@$!%*?&]{6,20}$/;
+
+    let inErrors = this.state.inErrors;
+    let emailValid = this.state.emailValid;
+    let passwordValid = this.state.passwordValid;
+    let confirmationValid = this.state.confirmationValid;
+
+    switch (fieldName) {
+      case 'email':
+
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Useless blank line.

File: app/src/containers/Signup/Signup.js [hunk 6]
```diff
@@ -1,9 +1,142 @@
...
+        emailValid = value.match(mail);
+        inErrors.email = emailValid ? '' : 'e-mail is not correct';
+        break;
+      case 'password':
+        passwordValid = value.match(pass);
+        inErrors.password = passwordValid ? '' : 'At least 6 characters, one (a-z), one (A-Z), one (0-9), one ($@$!%*?&)';
+
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Useless blank line.

File: app/src/containers/Signup/Signup.js [hunk 7]
```diff
@@ -1,9 +1,142 @@
...
+  }
+
+  errorClass = (error) => {
+    return (error.length === 0 ? '' : 'has-error');
+  }
+
+  handleButtonClick = (e) => {
+    e.preventDefault();
+    request
+    .post('http://localhost:8081/api/reg/user')
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: It should not be hardcoded.

File: app/src/containers/Signup/Signup.js [hunk 8]
```diff
@@ -1,9 +1,142 @@
...
+  });
+  }
+
+  render() {
     return (
-      <h1>Here will be signup page.</h1>
+      <div className='containerLog'>
+      <Form className='LoginForm' onSubmit={this.handleButtonClick}>
+          {['first_Name', 'last_Name'].map(param =>
+          <FieldGroup
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: You missed 2 spaces here.

File: app/src/containers/Signup/Signup.js [hunk 9]
```diff
@@ -1,9 +1,142 @@
...
+     }
+  });
+  }
+
+  render() {
     return (
-      <h1>Here will be signup page.</h1>
+      <div className='containerLog'>
+      <Form className='LoginForm' onSubmit={this.handleButtonClick}>
+          {['first_Name', 'last_Name'].map(param =>
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Use 2 spaces for indentation.

File: app/src/containers/Signup/Signup.scss [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+$marg: auto;
+$contain-width: 300px;
+$pad1: 10px;
+$pad2: 5px;
+
+
+.LoginForm{
+  width: $base-width;
+  margin:$marg;
+
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.scss]: Useless blank line.

File: app/src/containers/Signup/Signup.scss [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+$bor-width: 2px;
+$bor-color: white;
+$bck-color: #20B2AA;
+$marg: auto;
+$contain-width: 300px;
+$pad1: 10px;
+$pad2: 5px;
+
+
+.LoginForm{
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.scss]: Missed space before {

File: app/src/containers/Signup/Signup.scss [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  width: $base-width;
+  margin:$marg;
+  
+}
+
+.containerLog{
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.scss]: Missed space before {

File: app/src/containers/Signup/Signup.js [hunk 13]
```diff
@@ -1,9 +1,142 @@
...
+       return (browserHistory.push('/'), 'You have successfully signed up! For confirmation please visit your e-mail');
+     }
+  });
+  }
+
+  render() {
     return (
-      <h1>Here will be signup page.</h1>
+      <div className='containerLog'>
+      <Form className='LoginForm' onSubmit={this.handleButtonClick}>
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Fix indentation.

File: app/src/containers/Signup/Signup.js [hunk 14]
```diff
@@ -1,9 +1,142 @@
...
+    return (error.length === 0 ? '' : 'has-error');
+  }
+
+  handleButtonClick = (e) => {
+    e.preventDefault();
+    request
+    .post('http://localhost:8081/api/reg/user')
+    .set(user)
+    .end(function(err, res) {
+     if (err || !res.ok) {
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Fix indentation.

File: app/src/containers/Signup/Signup.js [hunk 15]
```diff
@@ -1,9 +1,142 @@
...
+    };
+  };
+
+  handleChange = (param) => (e) => {
+    let value = e.target.value;
+    this.setState({[param]: value},
+                  () => {this.validateField(param, value)});
+  };
+
+validateField = (fieldName, value) => {
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Missed space.

File: app/src/containers/Signup/Signup.js (hunk 1 above)
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Missed space before let.

File: app/src/containers/Signup/Signup.scss [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+	border: $bor-type $bor-width $bor-color;
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.scss]: Use 2 spaces.

File: app/src/containers/Signup/Signup.js (hunk 3 above)
Comment (avoytovich) [reply app/src/containers/Signup/Signup.js]: export default class Signup extends React.Component...чому не так?

File: app/src/containers/Signup/Signup.js [hunk 17]
```diff
@@ -1,9 +1,156 @@
...
+  lastName: 'Popov', 
+  email: 'rrr@ukr.net'
+};
+
+let FieldGroup = ({className, label, inErrors, ...props}) => (
+  <div>
+    <FormGroup className = {className}>
+      <ControlLabel>{label}</ControlLabel>
+      <FormControl {...props} />
+      {inErrors && <HelpBlock>{inErrors}</HelpBlock>}
```
Comment (avoytovich) [inline app/src/containers/Signup/Signup.js]: не можу зрозуміти, чому пишеш в JS як в JSX?

**Review by yarvanya — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

File: app/src/containers/Signup/Signup.js [hunk 18]
```diff
@@ -1,9 +1,156 @@
 import React from 'react';
+import { FormGroup, ControlLabel, FormControl, Button, Form, HelpBlock } from 'react-bootstrap';
+import { browserHistory } from 'react-router';
+import { URL } from './../../helper/constants';
+import request from 'superagent';
+import style from './Signup.scss';
+
+let user = {
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: hardcode ?

File: app/src/containers/Signup/Signup.scss [hunk 19]
```diff
@@ -0,0 +1,25 @@
+$base-width: 240px;
+$bor-radius: 25px;
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.scss]: Create common styles for all and use them and here add component styles

File: app/src/containers/Signup/Signup.js [hunk 20]
```diff
@@ -1,9 +1,156 @@
...
+    request
+    .post('URL.SIGNUP')
+    .set(user)
+    .end(function(err, res) {
+      if (err || !res.ok) {
+        return ('No connection with server');
+      } else {
+        return (
+          browserHistory.push('/'), 
+          'You have successfully signed up! For confirmation please visit your e-mail'
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: Add info messages file

File: app/src/containers/Signup/Signup.js [hunk 21]
```diff
@@ -1,9 +1,156 @@
...
+
+    switch (fieldName) {
+      case 'email':
+        emailValid = value.match(mail);
+        inErrors.email = emailValid ? '' : 'e-mail is not correct';
+        break;
+      case 'password':
+        passwordValid = value.match(pass);
+        inErrors.password = passwordValid ? '' : (
+            'At least 6 characters, one (a-z), one (A-Z), one (0-9), one ($@$!%*?&)'
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: Messages file

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/helper/constants.js [hunk 22]
```diff
@@ -0,0 +1,6 @@
+const API = {
+  HOST: 'http://localhost:',
+  PORT: '8082'
+}
+
+export default API;
```
Comment (yarvanya) [inline app/src/helper/constants.js]: Add an empty line...

File: app/src/common.scss [hunk 23]
```diff
@@ -0,0 +1,20 @@
+$border-type-solid: solid;
+$border-width-two: 2px;
+$border-color-green: green;
```
Comment (AlbaMontis) [inline app/src/common.scss]: Use hexadecimal colors or rgb.

File: app/src/containers/Signup/Signup.js [hunk 24]
```diff
@@ -1,9 +1,154 @@
...
+            type='submit'>Regist
+          </Button>
+        </Form>
+      </div>
     );
   }
 }
+
+const mapStatetoProps = (state, ownProps) => {
+  return {
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: With arrow functions you can return object without word "return": => ({val: prop})

File: app/src/redux/Signup.js [hunk 25]
```diff
@@ -0,0 +1,69 @@
+import request from 'superagent';
+import { browserHistory } from 'react-router';
```
Comment (AlbaMontis) [inline app/src/redux/Signup.js]: import { push } from 'react-router-redux';

File: app/src/redux/Signup.js [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+    .send(data)
+    .end(function(err, res) {
+      if (err || !res.ok) {
+        return (
+          dispatch(messageModal(messages.error)),
+          dispatch(showModal(true))
+        )
+      } else {
+        return (
+          browserHistory.push('/'),
```
Comment (AlbaMontis) [inline app/src/redux/Signup.js]: push('/');

File: app/src/redux/Signup.js [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+      return state;
+  }
+};
+
+export function showModal(show) {
+  return {
+    type: SHOW_MODAL,
+    show: show
+  };
+};
```
Comment (AlbaMontis) [inline app/src/redux/Signup.js]: > Semicolons should be included at the end of function expressions, but not at the end of function declarations.
https://google.github.io/styleguide/javascriptguide.xml#Semicolons

File: app/src/redux/Signup.js [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+  return {
+    type: MESSAGE_MODAL,
+    message: message
+  };
+};
+
+export function signupRequest() {
+  return {
+    type: SIGNUP_REQUEST
+};
```
Comment (AlbaMontis) [inline app/src/redux/Signup.js]: Missed space.

File: app/src/redux/Signup.js (hunk 26 above)
Comment (yurii-kokhalevych) [reply app/src/redux/Signup.js]: this must be ' ,'  with ';' doesn't work

File: app/src/redux/Signup.js (hunk 25 above)
Comment (yurii-kokhalevych) [reply app/src/redux/Signup.js]: it is not necessary

File: app/src/containers/Signup/Signup.js [hunk 29]
```diff
@@ -1,9 +1,150 @@
...
+  isValidate = () => {
+    this.setState({ 
+      Validate:
+        this.state.emailValid && 
+        this.state.passwordValid && 
+        this.state.confirmation === this.state.password
+    });
+  }
+
+  errorClass = (error) => {
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Useless ()

File: app/src/containers/Signup/Signup.js [hunk 30]
```diff
@@ -1,9 +1,150 @@
...
+    return (error.length === 0 ? '' : 'has-error');
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: You can write this arrow function without 'return'.

File: app/src/containers/Signup/Signup.js [hunk 31]
```diff
@@ -1,9 +1,150 @@
...
+      </div>
     );
   }
 }
+
+const mapStatetoProps = (state, ownProps) => ({
+  signup: state.signup
+});
+
+const mapDispatchToProps = (dispatch) => ({
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Useless ().

File: app/src/redux/Signup.js (hunk 26 above)
Comment (AlbaMontis) [reply app/src/redux/Signup.js]: I didn't mean ; , I meant react-router-redux method 'push'.

File: app/src/common.scss [hunk 32]
```diff
@@ -0,0 +1,20 @@
...
+$conteiner-width-modalMessage: 400px;
+$left-modalMessage: 30%;
+$top-modalMessage: 10%;
+$padding-thirty: 30px;
+$margin-auto: auto;
+$border-radius-five: 5px;
+$position-absolute: absolute;
+$base-width-registration: 250px;
+$border-radius-twentyFive: 25px;
+$border-color-white: #FFFFFF;
```
Comment (AlbaMontis) [inline app/src/common.scss]: Why not just white?

File: app/src/common.scss [hunk 33]
```diff
@@ -0,0 +1,20 @@
+$border-type-solid: solid;
+$border-width-two: 2px;
+$border-color-green: #008000;
+$border-color-red: #FF0000;
+$background-color-mint: #98ff98;
+$text-color-black: #000000;
```
Comment (AlbaMontis) [inline app/src/common.scss]: What if I need black somewhere else, not for text?

**Review by yarvanya — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

File: app/src/redux/Signup.js [hunk 34]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+
+export function signupUser(email, password) {
+  let data = {
+    email: email,
+    password: password
+  };
+
+return dispatch => {
```
Comment (AlbaMontis) [inline app/src/redux/Signup.js]: Missed space before return.

**Review by yarvanya — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

File: app/src/common.scss [hunk 35]
```diff
@@ -0,0 +1,9 @@
+$white: #ffffff;
+$green-dark: #008000;
+$green-ligth: #b3ffb3;
+$mint: #98ff98;
+$border-radius: 5px;
+$padding: 10px;
+$border-white: 2px solid $white;
+$border-green-dark: 2px solid $green-dark;
+$margin: auto;
```
Comment (4asof4ik) [inline app/src/common.scss]: drop

File: app/src/redux/reducer.js [hunk 36]
```diff
@@ -1,10 +1,12 @@
 import { combineReducers } from 'redux';
-import { routerReducer } from 'react-router-redux';
 import items from './Test';
+import signup from './Signup';
+import { routerReducer } from 'react-router-redux';
```
Comment (4asof4ik) [inline app/src/redux/reducer.js]: first should be vendor after that your

File: app/src/helper/messages.js [hunk 37]
```diff
@@ -0,0 +1,9 @@
+const messages = {
+  signup: 'You have successfully signed up! For confirmation please visit your e-mail',
```
Comment (4asof4ik) [inline app/src/helper/messages.js]: successSignup

File: app/src/helper/messages.js [hunk 38]
```diff
@@ -0,0 +1,9 @@
...
+  error: 'That email address is already in use',
```
Comment (4asof4ik) [inline app/src/helper/messages.js]: emailExistError

File: app/src/containers/Signup/Signup.js [hunk 39]
```diff
@@ -1,9 +1,138 @@
...
+    e.preventDefault();
+    this.props.actions.signupUser(this.state.email, this.state.password);
+  }
+
+  render() {
     return (
-      <h1>Here will be signup page.</h1>
+      <div className='containerLog'>
+        <Message />
+        <Form className='Signup' onSubmit={this.handleButtonClick}>
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: maybe use redux-form? https://redux-form.com/6.8.0/

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/containers/Signup/Signup.js (hunk 39 above)
Comment (yurii-kokhalevych) [reply app/src/containers/Signup/Signup.js]: We thought about redux-form earlier but decided use react-bootstrap

**Review by avoytovich — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

File: app/src/containers/Signup/Signup.js [hunk 40]
```diff
@@ -1,9 +1,138 @@
 import React from 'react';
+import { FormGroup, ControlLabel, FormControl, Button, Form, HelpBlock } from 'react-bootstrap';
+import { browserHistory } from 'react-router';
+import { Message } from '../../components';
+import messages from '../../helper/messages';
+import request from 'superagent';
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Do you need it here?

File: app/src/containers/Signup/Signup.js [hunk 41]
```diff
@@ -1,9 +1,138 @@
 import React from 'react';
+import { FormGroup, ControlLabel, FormControl, Button, Form, HelpBlock } from 'react-bootstrap';
+import { browserHistory } from 'react-router';
```
Comment (AlbaMontis) [inline app/src/containers/Signup/Signup.js]: Do you use it here?

**Review by AlbaMontis — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

File: app/src/containers/Signup/Signup.js [hunk 42]
```diff
@@ -1,9 +1,136 @@
...
+    let value = e.target.value;
+    this.setState({[param]: value},
+      () => {this.validateField(param, value);});
+  };
+
+  validateField = (fieldName, value) => {
+    const mail = /^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i;
+    const pass = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[$@$!%*?&])[A-Za-z\d$@$!%*?&]{6,20}$/;
+
+    let isErrors = this.state.isErrors;
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: let {isErrors, emailValid.............} = this.state;
check this

File: app/src/containers/Signup/Signup.js [hunk 43]
```diff
@@ -1,9 +1,136 @@
...
+        emailValid = value.match(mail);
+        isErrors.email = emailValid ? '' : `${messages.emailError}`;
+        break;
+      case 'password':
+        passwordValid = value.match(pass);
+        isErrors.password = passwordValid ? '' : `${messages.passwordError}`;
+        break;
+      case 'confirmation':
+        isErrors.confirmation = 
+          (this.state.confirmation === this.state.password) ? '' : `${messages.confirmationError}`;
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: read how separate ternary operator

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/redux/Signup.js [hunk 44]
```diff
@@ -0,0 +1,75 @@
...
+  return dispatch => {
+    dispatch(signupRequest());
+    request()
+      .post(API.HOST + API.PORT + '/api/users')
+      .send(data)
+      .end(function(err, res) {
+        if (err || !res.ok) {
+          if (err.status == 422) {
+            return (
+              dispatch(messageModal(messages.emailExistError)),
```
Comment (4asof4ik) [inline app/src/redux/Signup.js]: check spaces in this file

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/redux/Signup.js [hunk 45]
```diff
@@ -0,0 +1,75 @@
...
+    password: password
+  };
+
+  return dispatch => {
+    dispatch(signupRequest());
+    request()
+      .post(API.HOST + API.PORT + '/api/users')
+      .send(data)
+      .end(function(err, res) {
+        if (err || !res.ok) {
```
Comment (4asof4ik) [inline app/src/redux/Signup.js]: err || !res.ok && err.status == 422

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/redux/Signup.js (hunk 44 above)
Comment (yurii-kokhalevych) [reply app/src/redux/Signup.js]: this structure of spaces want travis

**Review by yarvanya — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

File: app/src/redux/Signup.js [hunk 46]
```diff
@@ -0,0 +1,68 @@
+import request from '../helper/request';
+import { push } from 'react-router-redux';
```
Comment (AlbaMontis) [inline app/src/redux/Signup.js]: Do you use it in this file?

**Review by avoytovich — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

File: app/src/containers/Signup/Signup.js [hunk 47]
```diff
@@ -0,0 +1,139 @@
+import React from 'react';
+import { FormGroup, ControlLabel, FormControl, Button, Form, HelpBlock } from 'react-bootstrap';
+import { Message } from '../../components';
+import messages from '../../helper/messages';
+import style from './Signup.scss';
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: import './Signup.scss';

File: app/src/containers/Signup/Signup.js [hunk 48]
```diff
@@ -0,0 +1,139 @@
...
+      () => {this.validateField(param, value);});
+  };
+
+  validateField = (fieldName, value) => {
+    const mail = /^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i;
+    const pass = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[$@$!%*?&])[A-Za-z\d$@$!%*?&]{6,20}$/;
+
+    let {isErrors, emailValid, passwordValid, confirmationValid} = this.state;
+
+    switch (fieldName) {
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: need rewrite

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

File: app/src/redux/Signup.js [hunk 49]
```diff
@@ -0,0 +1,67 @@
...
+export function signupUser(email, password) {
+  let data = {
+    email: email,
+    password: password
+  };
+
+  return dispatch => {
+    dispatch(signupRequest());
+    request()
+      .post(API.HOST + API.PORT + '/api/auth/users')
```
Comment (4asof4ik) [inline app/src/redux/Signup.js]: /api/auth/users this is not auth /api/users (fix in beck-end too)

File: app/src/containers/Signup/Signup.js [hunk 50]
```diff
@@ -0,0 +1,109 @@
...
+      email: /^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i,
+      password: /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[$@$!%*?&])[A-Za-z\d$@$!%*?&]{6,20}$/,
+      confirmation: this.state.password 
+    };
+    const valid = value.match(pattern[fieldName]);
+    const newState = this.state;
+    newState.isErrors[fieldName] = !valid ? messages[fieldName + 'Error'] : '';
+    const isEmpty = errors => {
+      let result = true;
+      for(let key in errors) {
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: use map

File: app/src/containers/Signup/Signup.js [hunk 51]
```diff
@@ -0,0 +1,109 @@
...
+    e.preventDefault();
+    this.props.actions.signupUser(this.state.email, this.state.password);
+  }
+
+  render() {
+    return (
+      <div className='containerLog'>
+        <Message />
+        <Form className='Signup' onSubmit={this.handleButtonClick}>
+          {['email', 'password', 'confirmation'].map(param =>
```
Comment (4asof4ik) [inline app/src/containers/Signup/Signup.js]: use the associative array with Object.keys

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: app/src/redux/Signup.js (hunk 49 above)
Comment (yurii-kokhalevych) [reply app/src/redux/Signup.js]: if I will not used auth I get error { success: false, message: 'No token provided.'}

**Review by yarvanya — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

File: app/src/redux/reducer.js [hunk 52]
```diff
@@ -1,5 +1,7 @@
 import { combineReducers } from 'redux';
 import { routerReducer } from 'react-router-redux';
+import items from './Test';
```
Comment (4asof4ik) [inline app/src/redux/reducer.js]: drop  test component and other (redux route .....)

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by yarvanya — DISMISSED** [review]

**Review by avoytovich — DISMISSED** [review]

**Review by AlbaMontis — DISMISSED** [review]

File: .babelrc [hunk 53]
```diff
@@ -1,3 +1,3 @@
 {
-  presets:["es2015", "stage-0", "react"],
+  presets:["es2015", "stage-0", "react"]
```
Comment (omaik) [inline .babelrc]: minor, whitespace after a colon.

**Review by yarvanya — APPROVED** [review]

**Review by avoytovich — APPROVED** [review]

**Review by AlbaMontis — APPROVED** [review]

**Review by 4asof4ik — APPROVED** [review]

**Pull request merged**